zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sStandardInterpreterLauncher.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.zeppelin.interpreter.recovery.RecoveryStorage;
+import org.apache.zeppelin.interpreter.remote.RemoteInterpreterUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Interpreter Launcher which use shell script to launch the interpreter process.
+ */
+public class K8sStandardInterpreterLauncher extends InterpreterLauncher {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(K8sStandardInterpreterLauncher.class);
+  private final Kubectl kubectl;
+  private InterpreterLaunchContext context;
+
+
+  public K8sStandardInterpreterLauncher(ZeppelinConfiguration zConf, RecoveryStorage recoveryStorage) throws IOException {
+    super(zConf, recoveryStorage);
+    kubectl = new Kubectl(zConf.getK8sKubectlCmd());
+    kubectl.setNamespace(getNamespace());
+  }
+
+  @VisibleForTesting
+  K8sStandardInterpreterLauncher(ZeppelinConfiguration zConf, RecoveryStorage recoveryStorage, Kubectl kubectl) {
+    super(zConf, recoveryStorage);
+    this.kubectl = kubectl;
+  }
+
+
+  /**
+   * Check if i'm running inside of kubernetes or not.
+   * It should return truth regardless of ZeppelinConfiguration.getRunMode().
+   *
+   * Normally, unless Zeppelin is running on Kubernetes, K8sStandardInterpreterLauncher shouldn't even have initialized.
+   * However, when ZeppelinConfiguration.getRunMode() is force 'k8s', InterpreterSetting.getLauncherPlugin() will try
+   * to use K8sStandardInterpreterLauncher. This is useful for development. It allows Zeppelin server running on your
+   * IDE and creates your interpreters in Kubernetes. So any code changes on Zeppelin server or kubernetes yaml spec
+   * can be applied without re-building docker image.
+   * @return
+   */
+  boolean isRunningOnKubernetes() {
+    if (new File("/var/run/secrets/kubernetes.io").exists()) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Get current namespace
+   * @throws IOException
+   */
+  String getNamespace() throws IOException {
+    if (isRunningOnKubernetes()) {
+      return readFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace", Charset.defaultCharset()).trim();
+    } else {
+      return "default";
+    }
+  }
+
+  /**
+   * Get hostname. It should be the same to Service name (and Pod name) of the Kubernetes
+   * @return
+   */
+  String getHostname() {
+    try {
+      return InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException e) {
+      return "localhost";
+    }
+  }
+
+  /**
+   * get Zeppelin server host dns.
+   * return <hostname>.<namespace>.svc.cluster.local
+   * @throws IOException
+   */
+  private String getZeppelinServiceHost() throws IOException {
+    if (isRunningOnKubernetes()) {
+      return String.format("%s.%s.svc.cluster.local",
+              getHostname(), // service name and pod name should be the same
```

Comment:
@Leemoonsoo Great work on making Zeppelin work in Kubernetes!

I have a concern with this approach of matching Service and Pod names via hostname. This will not work for zeppelin pods that are provisioned via Deployments or StatefulSets as pod names are dynamically generated during creation.

Reply:
Good point. Notebook serving (https://github.com/apache/zeppelin/pull/3356) has an exact problem while TestTask runs Zeppelin using `Job`, and ServingTask runs Zeppelin using `Deployment`.

In serving implementation, K8sStandardInterpreterLauncher has modified to looking for `SERVICE_NAME` env variable. Check https://github.com/apache/zeppelin/pull/3356/files#diff-dfad7c26921ec97aa7bca7fc3180408fL109. Does it make sense?

Reply:
That should work 👍